miner/src/main/java/fr/raksrinana/channelpointsminer/miner/database/BaseDatabase.java

```diff
@@ -67,14 +79,215 @@ public void addPrediction(@NotNull String channelId, @NotNull String eventId, @N
+		
+		try(var conn = getConnection();
+				var selectUserStatement = conn.prepareStatement("""
+						SELECT `ID` FROM `PredictionUser` WHERE `Username`=?""");
+				var predictionStatement = getPredictionStatement(conn)){
+			conn.setAutoCommit(false);
+			
+			username = username.toLowerCase(Locale.ROOT);
+			
+			selectUserStatement.setString(1, username);
+			var userResult = selectUserStatement.executeQuery();
+			int userId;
+			if(userResult.next()){
+				userId = userResult.getInt(1);
+			}
+			else{
+				try(var addUserStatement = conn.prepareStatement("""
+						INSERT INTO `PredictionUser`(`Username`) VALUES (?) RETURNING `ID`""")){
+					addUserStatement.setString(1, username);
+					var insertResult = addUserStatement.executeQuery();
+					insertResult.next();
+					userId = insertResult.getInt(1);
+				}
+				log.debug("Added new prediction user '{}'", username);
+			}
+			
+			predictionStatement.setInt(1, userId);
+			predictionStatement.setString(2, badge);
+			predictionStatement.setString(3, streamerName);
+			
+			predictionStatement.executeUpdate();
+			conn.commit();
+		}
+	}
+	
+	@NotNull
+	protected abstract PreparedStatement getPredictionStatement(@NotNull Connection conn) throws SQLException;
+	
+	@Override
+	public void cancelPrediction(@NotNull Event event) throws SQLException{
+		var ended = Optional.ofNullable(event.getEndedAt()).map(ZonedDateTime::toInstant).orElseGet(TimeFactory::now);
+		
+		try(var conn = getConnection();
+				var addCanceledPredictionStmt = conn.prepareStatement("""
+						INSERT INTO `ResolvedPrediction`(`EventID`,`ChannelID`, `Title`,`EventCreated`,`EventEnded`,`Canceled`)
+						VALUES (?,?,?,?,?,true)""");
+				var updateUserPredictionsStmt = conn.prepareStatement("""
+						UPDATE `UserPrediction`
+						SET `ResolvedPredictionID`=?
+						WHERE `ResolvedPredictionID`='' AND `ChannelID`=?""")
+		){
+			conn.setAutoCommit(false);
+			try{
+				//Add canceled prediction
+				addCanceledPredictionStmt.setString(1, event.getId());
+				addCanceledPredictionStmt.setString(2, event.getChannelId());
+				addCanceledPredictionStmt.setString(3, event.getTitle());
+				addCanceledPredictionStmt.setObject(4, event.getCreatedAt().withZoneSameInstant(UTC).toLocalDateTime());
+				addCanceledPredictionStmt.setObject(5, LocalDateTime.ofInstant(ended, UTC));
+				addCanceledPredictionStmt.executeUpdate();
+				
+				//Update made predictions with event-id
+				updateUserPredictionsStmt.setString(1, event.getId());
+				updateUserPredictionsStmt.setString(2, event.getChannelId());
+				updateUserPredictionsStmt.executeUpdate();
+				
+				conn.commit();
+			}
+			catch(SQLException e){
+				conn.rollback();
+				throw e;
+			}
+		}
+	}
+	
+	@Override
+	public void resolvePrediction(@NotNull Event event, @NotNull String outcome, @NotNull String badge, double returnRatioForWin) throws SQLException{
```

Comment:
Maybe would be a great idea to actually clear predictions of the channel instead of keeping them with the resolved id. Storing all choices of every user seems a bit much and may grow quite fast in size especially for large channels. What is interesting to us is the aggregated data.

Reply:
Added